### DISCUSSION TITLE
Synthetic Events via `/events` API

### DISCUSSION THREAD

**Discussion by aussig** [aussig/BGS-Tally#150] [Proposals]
BGS-Tally, through its internal logic, infers and tracks various activities that are missing direct game journal entries. All tallied activity is included as a snapshot via the `/activities` API but some of these activities could also be useful via the `/events` API to systems that consume the journal events stream.

This proposal is to introduce the concept of synthetic events - i.e. events delivered to the `/events` API that do not directly correspond with game journal entries. 

- Propose that all synthetic events have a unique prefix - `BGST_`? -  so as to:
    1. Avoid any potential conflict with future journal events introduced to the actual game 
    2. Clearly identify them as originating from BGS-Tally
- Propose the following new events:
    - `BGST_SpaceConflictZone` - Detail TBD
    - `BGST_SpaceConflictZoneObjective` - Detail TBD
    - `BGST_GroundConflictZone` - Detail TBD
    - `BGST_Scenario` - Detail TBD
 
Note that BGS-Tally's CZ tracking is still not perfect as the game does not log CZ wins, so various journal events are used to determine whether you have fought at a CZ.

Open question: Do we send a synthetic event when the user manually tallies a CZ via the UI?  What happens if the user 'untallies' (decrements) the CZs count?

**Comment by cluster-fox** [comment]
Here is my response:

The logic for synthetic events is that some event tracking is a LOT more efficient if done on the client side.

The best example is Conflict Zones. Uploading all related events and having them processed server side is a lot of overhead and introduces many more possible points of failure while the client has local access to the journal and can send out a single event rather than multiple journal entries.

I suggest the prefix `Synthetic_` or `BGS_` as the prefix should not be tied to a client, but defined in the API : 
- `Synthetic_ConflictZone` (with attributes for different types)
- `Synthetic_GroundConflictZone` (with attributes for data, such as the settlement)
- `Synthetic_Scenario`

In my opinion, `Synthetic_ConflictZoneObjective` should be part of the `Synthetic_ConflictZone` event since they are 100% related.

I agree about different Events for Space and Ground since those are not that much related and have different effects.

--
About the perfection of tracking. It's not expected that any client will be perfect since we don't have direct access to the CZ completion events. Being 80%+ there is enough. Precision is not really required since from a BGS perpective, it's impossible to know exactly how much the opposition is doing.

--
As for the open question. In my opinion, manual user entry should trigger events that account for +1 or -1. Should that get out of sync somehow, the `/activities` endpoint can be used to validate.

**Comment by aussig** [comment]
I agree the prefix should be non client specific, and `Synthetic_` LGTM. 

Not so sure about combining the main event and objectives. At the moment we track a CZ once a) the user has dropped into one and b) they have made a first kill. There is a [separate proposal](https://github.com/aussig/BGS-Tally/discussions/151) to change this to include kill counting, but that also comes with downsides as outlined in that proposal. 

Either way, objectives are completed at a different time to the main CZ. I would argue that as `/events` is a stream, they should be triggered at the time of detection, as if they had appeared in the journal at that time. 

So for manual user entry, if you would be happy with the concept of decrementing via a “negative count” mechanism, we could achieve that with a separate event (bit clumsy and introduces event bloat) or with a `count` parameter on the main events (neater IMHO) where `count` could be `1` (default) or `-1`

**Reply by cluster-fox** [reply]
In this case events are streamed if detected but can be bunched up if entered/corrected manually. So yes, time of detection OR manually.

We can design this at the API level but for maximum flexibility, I'm a strong proponent of `low = INT, med = INT, high = INT` all optional parameters. This format carries all required meaning, is very short and gives maximum flexibility for the use case.

GroundCZ simply needs an additional field `settlement = STR`. This is fine because stream events would catch the settlement while manual entry would likely not but even if they do, `low=1, settlement="Name"` or `low=3, settlement="Name"` are both totally acceptable. Again, can be designed at the API level.

You make a good case for the sideObjective being separate.

**Comment by cluster-fox** [comment]
I got a proposal in the API definition. Didn't make the Synthetic_CZObjective or Synthetic_Scenario yet but I'd stick with roughly the same format. Those types Could include an optional type.

**Comment by aussig** [comment]
Implementation in issue #288
